pkg/flavor/flavor.go

```diff
@@ -142,3 +244,398 @@ func AddFlavorNameAnnotations(vm *virtv1.VirtualMachine, target metav1.Object) {
+	}
+
+	if flavorSpec.Memory.Hugepages != nil {
+		flavorHugePages := flavorSpec.Memory.Hugepages.DeepCopy()
+		vmiSpec.Domain.Memory.Hugepages = flavorHugePages
+	}
+
+	return nil
+}
+
+func applyIOThreadPolicy(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
+
+	if flavorSpec.IOThreadsPolicy == nil {
+		return nil
+	}
+
+	if vmiSpec.Domain.IOThreadsPolicy != nil {
+		return Conflicts{field.Child("domain", "ioThreadsPolicy")}
+	}
+
+	flavorIOThreadPolicy := *flavorSpec.IOThreadsPolicy
+	vmiSpec.Domain.IOThreadsPolicy = &flavorIOThreadPolicy
+
+	return nil
+}
+
+func applyLaunchSecurity(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
+
+	if flavorSpec.LaunchSecurity == nil {
+		return nil
+	}
+
+	if vmiSpec.Domain.LaunchSecurity != nil {
+		return Conflicts{field.Child("domain", "launchSecurity")}
+	}
+
+	vmiSpec.Domain.LaunchSecurity = flavorSpec.LaunchSecurity.DeepCopy()
+
+	return nil
+}
+
+func applyGPUs(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
+
+	if len(flavorSpec.GPUs) == 0 {
+		return nil
+	}
+
+	if len(vmiSpec.Domain.Devices.GPUs) >= 1 {
+		return Conflicts{field.Child("domain", "devices", "gpus")}
+	}
+
+	vmiSpec.Domain.Devices.GPUs = make([]v1.GPU, len(flavorSpec.GPUs))
+	copy(vmiSpec.Domain.Devices.GPUs, flavorSpec.GPUs)
+
+	return nil
+}
+
+func applyHostDevices(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
+
+	if len(flavorSpec.HostDevices) == 0 {
+		return nil
+	}
+
+	if len(vmiSpec.Domain.Devices.HostDevices) >= 1 {
+		return Conflicts{field.Child("domain", "devices", "hostDevices")}
+	}
+
+	vmiSpec.Domain.Devices.HostDevices = make([]v1.HostDevice, len(flavorSpec.HostDevices))
+	copy(vmiSpec.Domain.Devices.HostDevices, flavorSpec.HostDevices)
+
+	return nil
+}
+
+func applyDevicePreferences(preferenceSpec *flavorv1alpha1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
+
+	if preferenceSpec.Devices == nil {
+		return
+	}
+
+	// We only want to apply a preference bool when...
+	//
+	// 1. A preference has actually been provided
+	// 2. The preference is true
+	// 3. The user hasn't defined the corresponding attribute already within the VMI
+	//
+	if preferenceSpec.Devices.PreferredAutoattachGraphicsDevice != nil && *preferenceSpec.Devices.PreferredAutoattachGraphicsDevice && vmiSpec.Domain.Devices.AutoattachGraphicsDevice == nil {
+		vmiSpec.Domain.Devices.AutoattachGraphicsDevice = pointer.BoolPtr(true)
+	}
+
+	if preferenceSpec.Devices.PreferredAutoattachMemBalloon != nil && *preferenceSpec.Devices.PreferredAutoattachMemBalloon && vmiSpec.Domain.Devices.AutoattachMemBalloon == nil {
+		vmiSpec.Domain.Devices.AutoattachMemBalloon = pointer.BoolPtr(true)
+	}
```

Comment:
nit: this is very verbose.. maybe we can use a helper function to clean things up a bit?
on second thought, it's not trivial here since you would need to provide both `preferenceSpec.Devices.PreferredAutoattachMemBalloon` and `vmiSpec.Domain.Devices.AutoattachMemBalloon` to this helper function..

don't know, but worth thinking about :)

Reply:
Yeah I wanted to go back and refactor all of this a little but wanted to get to the end of the series first. A helper should be trivial enough to do, I'll try to rework this today thanks!